src/components/structures/RoomView.tsx

```diff
@@ -621,61 +574,48 @@ export class RoomView extends React.Component<IRoomProps, IRoomState> {
-                this.setState({ showDisplaynameChanges: value as boolean }),
-            ),
-        ]);
+        if (roomId !== undefined) {
+            this.settingWatchers.push(
+                SettingsStore.watchSetting("showReadReceipts", roomId, (...[, , , value]) =>
+                    this.setState({ showReadReceipts: value as boolean }),
+                ),
+                SettingsStore.watchSetting("showRedactions", roomId, (...[, , , value]) =>
+                    this.setState({ showRedactions: value as boolean }),
+                ),
+                SettingsStore.watchSetting("showJoinLeaves", roomId, (...[, , , value]) =>
+                    this.setState({ showJoinLeaves: value as boolean }),
+                ),
+                SettingsStore.watchSetting("showAvatarChanges", roomId, (...[, , , value]) =>
+                    this.setState({ showAvatarChanges: value as boolean }),
+                ),
+                SettingsStore.watchSetting("showDisplaynameChanges", roomId, (...[, , , value]) =>
+                    this.setState({ showDisplaynameChanges: value as boolean }),
+                ),
+            );
+        }
 
         if (!initial && this.state.shouldPeek && !newState.shouldPeek) {
             // Stop peeking because we have joined this room now
-            this.context.client.stopPeeking();
-        }
-
-        // Temporary logging to diagnose https://github.com/vector-im/element-web/issues/4307
-        logger.log(
-            "RVS update:",
-            newState.roomId,
-            newState.roomAlias,
-            "loading?",
-            newState.roomLoading,
-            "joining?",
-            newState.joining,
-            "initial?",
-            initial,
-            "shouldPeek?",
-            newState.shouldPeek,
-        );
+            client.stopPeeking();
+        }
 
         // NB: This does assume that the roomID will not change for the lifetime of
         // the RoomView instance
         if (initial) {
-            newState.room = this.context.client!.getRoom(newState.roomId) || undefined;
+            newState.room = client.getRoom(newState.roomId) || undefined;
             if (newState.room) {
                 newState.showApps = this.shouldShowApps(newState.room);
                 this.onRoomLoaded(newState.room);
             }
         }
 
-        if (this.state.roomId === null && newState.roomId !== null) {
+        if (this.state.roomId === undefined && newState.roomId !== undefined) {
             // Get the scroll state for the new room
 
             // If an event ID wasn't specified, default to the one saved for this room
             // in the scroll state store. Assume initialEventPixelOffset should be set.
             if (!newState.initialEventId) {
-                const roomScrollState = RoomScrollStateStore.getScrollState(newState.roomId);
+                const roomScrollState = RoomScrollStateStore!.getScrollState(newState.roomId);
```

Comment:
We should probably use optional chaining here

```suggestion
                const roomScrollState = RoomScrollStateStore?.getScrollState(newState.roomId);
```

Reply:
This is kind of a weird one. In reality, `RoomScrollStateStore` is guaranteed to be defined:
https://github.com/matrix-org/matrix-react-sdk/blob/e9d723269f2e503a2c0ca6895c55310280dd73e4/src/stores/RoomScrollStateStore.ts#L50-L53
I believe this is a pre-TS artifact, which is why the default export is *technically* `RoomScrollStateStore | undefined`. Truthfully, the assertion belongs in that file (since it guarantees its presence).